core/shared/src/main/scala/org/http4s/Entity.scala

```diff
@@ -19,20 +19,45 @@ package org.http4s
 import cats.Monoid
 import cats.syntax.all._
 import cats.~>
+import fs2.Pure
+import fs2.Stream
 
-final case class Entity[+F[_]](body: EntityBody[F], length: Option[Long] = None) {
-  def translate[F1[x] >: F[x], G[_]](fk: F1 ~> G): Entity[G] = Entity(body.translate(fk), length)
+sealed trait Entity[+F[_]] {
+  def body: EntityBody[F]
+  def length: Option[Long]
```

Comment:
As I pointed in #5754 - would this better to have the `length` as an effectful function?  I'm confused a bit with a case when `Response` will have a non-empty `Stream` of data and empty `length`.

Reply:
An empty length actually means unknown length. I.e., the only way to compute the length is by evaluating the entire `Stream`, which you might not want to do. Maybe we just need a scaladoc to explain this? :)

Reply:
Maybe we should rename `length` to `knownLength`? It's a bit awkward though

Reply:
@armanbilge are we need that `length` then? I think it could be used in some surprising ways on the user site and can bring nasty bugs for those who come to `http4s` with similar to `akka-http` experience (where data is a fixed array of bytes).  I'm wondering about something like that ADT:
```
Entity = Empty | EntityWithLength | DefaulEntity
```
What do you folks think?

Reply:
> `akka-http` experience (where data is a fixed array of bytes)

Can you explain more about this? I'm looking at akka-http scaladocs for `HttpEntity`:

https://doc.akka.io/api/akka-http/10.2.7/akka/http/scaladsl/model/HttpEntity.html

All I see is a `Stream` of bytes:
```scala
 def dataBytes: Source[ByteString, Any]
```

They also have a method similar to the one that concerns you:
```scala
/** Some(content length) if a length is defined for this entity, None otherwise.
 * A length is only defined for Strict and Default entity types.
 */
abstract def contentLengthOption: Option[Long]
```

So, I think we can provide a similar experience. What do you think?

Reply:
I mean that it's quite common in the `akka-http` land to use a strict version of the `HttpEntity` - `HttpEntity.Strict`. `contentLengthOption` is a not good design choice, I think. Even `akka-http` itself contains warnings about using it:
```
In many cases it's dangerous to rely on the (non-)existence of a content-length.
HTTP intermediaries like (transparent) proxies are allowed to change the transfer-encoding which can result in the entity being delivered as another type as expected.
```
So what I stand for - it's a safer API for end-users. Since we are trying to design a model of `HttpEntity` from scratch, we can offer to users something more robust than scaladoc with warnings, I opine.

Reply:
Ah, thanks for explaining :)

First of all, 💯 👍 to some sort of `StrictEntity` in our `Entity` ADT. That is definitely one of the goals of these changes, that I think (hope) everyone wants 😆 

Regarding `length: Option[Long]` you make an important point there. It seems there is an important distinction whether it is derived from the `Content-Length` header or from the actual known size of a strict entity in memory. The latter should be safe, and maybe in any other case it should be `None`. Curious to hear others' thoughts on this.

Reply:
Yep, `Strict` is coming up next! That will wrap a `Chunk[byte]` and have length defined as the size of the chunk.

Streaming bodies are the harder case I think - I haven't yet properly digested how Chris's PR handles them, but there's a `Chunked` case that always has length None and a `TrustMe` case where length is set by calling code - maybe we can more discuss these issues more usefully and concretely when we get to those slices?

I still don't see how an effectful `length` field would work with an immutable data model though - computing the size would require running the body stream, leaving it in a bad state for actually running the body unless the body is memoized. (What we *could* do is to add a `prefetch` method that effectfully converts streaming entities to strict ones.)

Reply:
Sorry, I made the same error twice. I thought this design of `HttpEntity` is the final. Probably I should dig into Chris PR :) 
And yeah, if other slices are coming, we can discuss those things there.